src/lib/PerceptronPredictor.cc

```diff
@@ -0,0 +1,153 @@
+#include "simeng/PerceptronPredictor.hh"
+
+namespace simeng {
+
+PerceptronPredictor::PerceptronPredictor(ryml::ConstNodeRef config)
+    : btbBits_(config["Branch-Predictor"]["BTB-Tag-Bits"].as<uint64_t>()),
+      globalHistoryLength_(
+          config["Branch-Predictor"]["Global-History-Length"].as<uint64_t>()),
+      rasSize_(config["Branch-Predictor"]["RAS-entries"].as<uint64_t>()) {
+  // Build BTB based on config options
+  uint32_t btbSize = (1 << btbBits_);
+  btb_.resize(btbSize);
+  for (int i = 0; i < btbSize; i++) {
+    btb_[i].first.assign(globalHistoryLength_, 0);
+    btb_[i].first.push_back(1);
+    btb_[i].second = 0;
+  }
+
+  // Set up training threshold according to empirically determined formula
+  trainingThreshold_ = (int)((1.93 * (globalHistoryLength_)) + 14);
+}
+
+PerceptronPredictor::~PerceptronPredictor() {
+  ras_.clear();
+  rasHistory_.clear();
+}
+
+BranchPrediction PerceptronPredictor::predict(uint64_t address, BranchType type,
+                                              int64_t knownOffset) {
+  // Get index via an XOR hash between the global history and the lower btbBits_
+  // bits of the instruction address
+  uint64_t hashedIndex = ((address >> 2) ^ globalHistory_) & ((1 << btbBits_) - 1);
```

Comment:
This comment and subsequent line has confused me a bit. Not sure if "XOR hash" is a phrase (may well be) could replace with just "XOR". "btbBits_" could be quoted e.g. 'btbBits_', to make it clear that it is the _number_ of lower bits. If I have then interpreted the comment correctly I don't think it aligns with the code which is first XOR the address with the history then take the lower bits as opposed to what the comment suggests which would be (address & lower bits) ^ globalHistory_. These are not logically equivalent, but you probably get away with it as it is deterministic. 

With a little bit more thinking I reckon your code is correct but the comment isn't. You want the hash from the xor between the history and the address but you don't want to use this as an index as it could be out of range of the btb_ so you take the lower bits to ensure it is within the bounds. Is that the idea here?